subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformIntegrationTest.groovy

```diff
@@ -1594,6 +1595,46 @@ Found the following transforms:
         failure.assertHasCause("Transform output file ${testDirectory.file('other.jar')} is not a child of the transform's input file or output directory.")
     }
 
+    def "user gets a reasonable error message when transform registers an output that is not in an input or in the output directory"() {
+        given:
+        buildFile << """
+            def a = file('a.jar')
+            a.text = '1234'
+
+            dependencies {
+                compile files(a)
+            }
+
+            SomewhereElseTransform.output = file("other.jar")
+
+            class SomewhereElseTransform implements ArtifactTransformAction {
+                static def output
+                void transform(ArtifactTransformOutputs outputs) {
+                    def outputFile = outputs.file(output)
+                    outputFile.text = "123"
+                }
+            }
+            ${declareTransformAction('SomewhereElseTransform')}
+
+            task resolve(type: Copy) {
+                def artifacts = configurations.compile.incoming.artifactView {
+                    attributes { it.attribute(artifactType, 'size') }
+                }.artifacts
+                from artifacts.artifactFiles
+                into "\${buildDir}/libs"
+            }
+        """
+
+        when:
+        fails "resolve"
+
+        then:
+        failure.assertHasDescription("Execution failed for task ':resolve'.")
+        failure.assertHasCause("Could not resolve all files for configuration ':compile'.")
+        failure.assertHasCause("Failed to transform file 'a.jar' to match attributes {artifactType=size}")
+        failure.assertHasCause("Transform output file ${testDirectory.file('other.jar')} is not a child of the transform's input file or output directory.")
```

Comment:
The input artifact can be a directory, which it is in most cases for filtering transforms. Think unzip and then select some files in the unzipped dir.

Reply:
I'm not sure what you are replying to here, but I agree fully. :) In any case, if you are referring to the first sentence, then what I meant was a _file_ can't have descendants, whereas a _directory_ can.

Reply:
> A `file` can't have descendants, and we should probably find a more specific word to replace "child". And `transform's input file` should be called the "input artifact" now, right?

How about this: 
```
Transform output ${testDirectory.file('other.jar')} is not part of the transform's inputs nor in the transform's workspace.
```

Reply:
Better, but I think we can still improve it:

* we never define the `transform's workspace` either, so I'm not sure if this is a good way to capture the problem
* it's not really the transform's _inputs_ that this should be part of, e.g. you can't return a file from among the transform's secondary inputs, or the dependencies of the input artifact; instead the registered file should be part of the input artifact itself.
* the phrasing sounds passive-aggressive (it's not unique to this error message by far) — instead of just saying that "something didn't have quality X" and then fail, we should say that "something must have quality X, but did not".

Reply:
I ended up with this message:
```
Transform output 'other.jar' must be a part of the input artifact or a relative path.
```

WDYT?

Reply:
`... or refer to a relative path`? Without the verb it's a bit confusing.

Reply:
Changed to what you suggested.